### PR TITLE
Add Cube Semantic Layer Agent

### PR DESCRIPTION
**Pull request by johnmai-dev**
- **Description:** SQL Chains and Agents support building and running SQL queries based on natural language prompts. However, when faced with complex queries, they may not perform well in producing results.
Now, I'm providing a Cube Semantic Layer Agent, which can better support complex queries. A Cube data model is used to transform raw data into meaningful business definitions and pre-aggregate data for optimal results.
  - **Twitter handle:** @JohnMai95
  - **Tag maintainer:** @baskaryan @hwchase17 @eyurtsev 
 
cube_semantic_layer_agent

ag

### CODE REVIEW

**Comment by baskaryan** [comment]
This looks awesome! Are there any security concerns with executing LLM generated cube queries? if so would prefer to move the chain and agent to langchain_experimental and add some security warnings to the docstring and docs

cc @MikeNitsenko @eyurtsev

**Comment by maiqingqiang** [comment]
> This looks awesome! Are there any security concerns with executing LLM generated cube queries? if so would prefer to move the chain and agent to langchain_experimental and add some security warnings to the docstring and docs
> 
> cc @MikeNitsenko @eyurtsev

It is not susceptible to SQL Injection and is not affected by other security issues.
- The queries can only utilize metadata defined within the cube. If the LLM generates non-existent metadata, data retrieval will not be possible.
- The Cube does not support  created, modified, or deleted operations.

**Comment by eyurtsev** [comment]
Hi @maiqingqiang thank you for the contribution! Code quality is great, so we can definitely merge.

Before the merge we need to add in code documentation + security note. Documentation should go for the agent/chain/tools -- OK to duplicate. 

Things to document:
* Uses rest API to interact with the data
* Include link to API spec:  https://cube.dev/docs/reference/rest-api
* Explain the API does (e.g., offers read only queries over the data -- if that's true).
* Say in the security note that that a user can potentially access any of the data that's avaiable

Is there any row level security provided for the data? For example, if I were to upload an orders tables with customer ids, is there a way to safely run queries on behalf of a particular customer (without seeing any other customer data?)


nits: 
* Could you update the code to use `*` prior to any named arguments?
* If you're able to provide any unit-tests that would be appreciated, but is not required.

**Comment by maiqingqiang** [comment]
> Hi @maiqingqiang thank you for the contribution! Code quality is great, so we can definitely merge.
> 
> Before the merge we need to add in code documentation + security note. Documentation should go for the agent/chain/tools -- OK to duplicate.
> 
> Things to document:
> 
> * Uses rest API to interact with the data
> * Include link to API spec:  https://cube.dev/docs/reference/rest-api
> * Explain the API does (e.g., offers read only queries over the data -- if that's true).
> * Say in the security note that that a user can potentially access any of the data that's avaiable
> 
> Is there any row level security provided for the data? For example, if I were to upload an orders tables with customer ids, is there a way to safely run queries on behalf of a particular customer (without seeing any other customer data?)
> 
> nits:
> 
> * Could you update the code to use `*` prior to any named arguments?
> * If you're able to provide any unit-tests that would be appreciated, but is not required.

Updated:
- Use `*` prior to any named arguments
- Remove something whitespace
- Add security note

Todo:
- docs
- unit tests
I will update the docs and unit tests later.

About row level security:
Cube's official website provides Tips and Tricks.
https://cube.dev/docs/product/faqs/tips-and-tricks#can-i-bypass-row-level-security-when-using-the-sql-api

**Comment by maiqingqiang** [comment]
Hello, @paveltiunov @keydunov. I have some questions related to Cube security. Could you please provide some insights or recommendations?

**Comment by maiqingqiang** [comment]
Hi @eyurtsev I have completed the modifications as per your feedback.

**Comment by maiqingqiang** [comment]
Could you please assist me by reviewing this? @eyurtsev @baskaryan

**Comment by maiqingqiang** [comment]
cc 🫶🏻 @MikeNitsenko @eyurtsev @baskaryan

**Comment by MikeNitsenko** [comment]
> cc 🫶🏻 @MikeNitsenko @eyurtsev @baskaryan

Will try to review it today. Thanks!

File: libs/langchain/langchain/utilities/__init__.py
Comment (MikeNitsenko) [inline libs/langchain/langchain/utilities/__init__.py:334]: Please consider using `CubeAPIWrapper` name for consistency.

File: docs/docs/use_cases/qa_structured/cube.ipynb
Comment (MikeNitsenko) [inline docs/docs/use_cases/qa_structured/cube.ipynb:301]: Could you please explain why would you want to utilize `/sql` endpoint?

File: libs/langchain/langchain/agents/agent_toolkits/cube/prompt.py
Comment (MikeNitsenko) [inline libs/langchain/langchain/agents/agent_toolkits/cube/prompt.py:11]: Please consider using `Cube Semantic Layer` over `Cube Semantic` for consistency

File: libs/langchain/langchain/chains/cube/prompt.py
Comment (MikeNitsenko) [inline libs/langchain/langchain/chains/cube/prompt.py:8]: > You can order the results by a relevant column to return the most interesting examples in the database.

Maybe it is worth to use more specific word than `interesting` to explain LLM its final goal.
Probably replacing `interesting` with `relevant` could be more evident to LLM despite duplicating the words.

File: libs/langchain/langchain/utilities/__init__.py
Comment (MikeNitsenko) [inline libs/langchain/langchain/utilities/__init__.py:291]: Consider using `CubeAPIWrapper` for consistency

File: libs/langchain/langchain/utilities/cube.py
Comment (MikeNitsenko) [inline libs/langchain/langchain/utilities/cube.py:338]: Could you please explain why do you need the `/sql` endpoint?

**Review by MikeNitsenko — COMMENTED** [review]
Great job, @maiqingqiang !
Looks good to me, please take a look at comments.

File: libs/langchain/langchain/utilities/cube.py
Comment (maiqingqiang) [reply libs/langchain/langchain/utilities/cube.py:338]: Sometimes, we do not wish to directly query data; we simply want to obtain the SQL. Subsequently, there might be a need to make secondary modifications to the SQL based on certain business requirements. That's why we are adding a /sql endpoint.

File: docs/docs/use_cases/qa_structured/cube.ipynb
Comment (maiqingqiang) [reply docs/docs/use_cases/qa_structured/cube.ipynb:303]: Sometimes, we do not wish to directly query data; we simply want to obtain the SQL. Subsequently, there might be a need to make secondary modifications to the SQL based on certain business requirements. That's why we are adding a /sql endpoint.

**Comment by maiqingqiang** [comment]
Hi @MikeNitsenko  I have completed the modifications as per your feedback.

**Review by MikeNitsenko — APPROVED** [review]
LGTM

**Comment by maiqingqiang** [comment]
cc @baskaryan 🫶🏻

**Comment by maiqingqiang** [comment]
Are there any other issues currently blocking the merge?  @baskaryan 🫶🏻

**Comment by eyurtsev** [comment]
@MikeNitsenko / @maiqingqiang apologies we never got around to merging this -- we are to blame and apologies for all the effort you've spent so far. 

If you're still interested in contributing this integration, please break it into smaller PRs and try to PR the tools / utilities in first. 

At the moment, we'll likely not merge a custom agent implementation into the code, we're trying to move away from this pattern to put more of the configuration into user code.

I'm going to close this PR for now. Feel free to re-open or reach out if you want to discuss what would make sense (can reach out via email eugene at langchain.dev)

**Comment by maiqingqiang** [comment]
It's so frustrating, waited for more than half a year.😓

**Pull request closed without merging**
